packages/frontend/component/src/hooks/use-disposable.ts

```diff
@@ -0,0 +1,60 @@
+import { useEffect, useState } from 'react';
+
+export function useDisposable<T extends Disposable | AsyncDisposable>(
+  disposableFn: (abortSignal?: AbortSignal) => Promise<T | null>,
+  deps?: any[]
+): { data: T | null; loading: boolean; error: Error | null };
+
+export function useDisposable<T extends Disposable | AsyncDisposable>(
+  disposableFn: (abortSignal?: AbortSignal) => T | null,
+  deps?: any[]
+): { data: T | null };
+
+export function useDisposable<T extends Disposable | AsyncDisposable>(
+  disposableFn: (abortSignal?: AbortSignal) => Promise<T | null> | T | null,
+  deps?: any[]
+) {
+  const [state, setState] = useState<{
+    data: T | null;
+    loading: boolean;
+    error: Error | null;
+  }>({
+    data: null,
+    loading: false,
+    error: null,
+  });
+
+  useEffect(() => {
+    const abortController = new AbortController();
+    let _data: T | null = null;
+    setState(prev => ({ ...prev, loading: true, error: null }));
+
+    Promise.resolve(disposableFn(abortController.signal))
+      .then(data => {
+        _data = data;
+        if (!abortController.signal.aborted) {
+          setState({ data, loading: false, error: null });
+        }
+      })
+      .catch(error => {
+        if (!abortController.signal.aborted) {
+          setState(prev => ({ ...prev, error, loading: false }));
+        }
+      });
+
+    return () => {
+      abortController.abort();
+
+      if (_data && typeof _data === 'object') {
+        if (Symbol.dispose in _data) {
+          _data[Symbol.dispose]();
+        } else if (Symbol.asyncDispose in _data) {
+          _data[Symbol.asyncDispose]();
+        }
+      }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [...(deps || [])]);
```

Comment:
```
deps || []
```?

Reply:
`deps = [x, y, z]` in `useXxx(fn, [x, y, z])`. The deps should be destructed here. Otherwise we need to memo `[x, y, z]`.

Reply:
I don't get it. array passed to react as deps will be shallow compared element by element

Reply:
im sorry, my bad. the wrapper brackets should be removed

Reply:
I get what you mean. yeah, we may not need to wrap the value with brackets